src/snovault/schema_utils.py

```diff
@@ -60,6 +60,55 @@ def mixinProperties(schema, resolver):
     return schema
 
 
+def resolve_ref(ref, resolver):
+    with resolver.resolving(ref) as resolved:
+        if not isinstance(resolved, dict):
+            raise ValueError(
+                f'Schema ref {ref} must resolve dict, not {type(resolved)}'
+            )
+        return resolved
+
+
+def resolve_refs(data, resolver):
+    if isinstance(data, dict):
+        # Return copy.
+        resolved_data = {}
+        for k, v in data.items():
+            if k == '$ref':
+                # Assumes resolved value is dictionary.
+                resolved_data.update(
+                    # Recurse here in case the resolved value has refs.
+                    resolve_refs(
+                        # Actually get the ref value.
+                        resolve_ref(v, resolver),
+                        resolver
+                    )
+                )
+            else:
+                resolved_data[k] = resolve_refs(v, resolver)
```

Comment:
I don't think this behaviour matches either the current or older specs (see my other comment). To match the drafts 4-7 behaviour it would need to look something like:
```py
        if k == '$ref':
            # Assumes resolved value is dictionary.
              resolved_data.update(
                  # Recurse here in case the resolved value has refs.
                  resolve_refs(
                      # Actually get the ref value.
                      resolve_ref(v, resolver),
                      resolver
                  )
              )
        else:
            for k, v in data.items():
                resolved_data[k] = resolve_refs(v, resolver)
```